File: deck.py
from card import Card
import random

class Deck:
    def __init__(self):
        self.suits = ["hearts", "clubs", "spades", "diamonds"]
        self.numbers = [2, 3, 4, 5, 6, 7, 8, 9, 10, "Ace", "Jack", "King", "Queen"]
        self.deck = []
        self.create_deck()
        self.shuffle_deck()

    def create_deck(self):
        for suit in self.suits:
            for number in self.numbers:
                current_card = Card(suit, number)
                self.deck.append(current_card)

    def shuffle_deck(self):
        random.shuffle(self.deck)

    def deal_card(self):
        return self.deck.pop()

    def display_deck(self):
        for card in self.deck:
            card.display()

if __name__ == "__main__":
    deck_instance = Deck()
    deck_instance.display_deck()

File: blackjack-sim/src/components/Deck.ts
import Card from "./Card";

class Deck {
  suits: string[];
  numbers: (number | string)[];
  deck: Card[];
  readonly initialSize: number;

  constructor(numDecks: number = 1) {
    this.suits = ["hearts", "clubs", "spades", "diamonds"];
    this.numbers = [2, 3, 4, 5, 6, 7, 8, 9, 10, "Ace", "Jack", "King", "Queen"];
    this.deck = [];
    this.createDeck(numDecks);
    this.initialSize = this.deck.length;
    this.shuffleDeck();
  }

  createDeck(numDecks: number): void {
    for (let i = 0; i < numDecks; i++) {
      for (const suit of this.suits) {
        for (const number of this.numbers) {
          this.deck.push(new Card(suit, number.toString()));
        }
      }
    }
  }

  /** Fisher–Yates shuffle (uniform permutation). */
  shuffleDeck(): void {
    for (let i = this.deck.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.deck[i], this.deck[j]] = [this.deck[j], this.deck[i]];
    }
  }

  dealCard(): Card | undefined {
    return this.deck.pop();
  }

  cardsRemaining(): number {
    return this.deck.length;
  }

  displayDeck(): void {
    this.deck.forEach((card) => {
      console.log(card.toString());
    });
  }
}

export default Deck;
